reactive/kotlinx-coroutines-rx2/src/RxScheduler.kt

```diff
@@ -4,16 +4,118 @@
 
 package kotlinx.coroutines.rx2
 
-import io.reactivex.Scheduler
+import io.reactivex.*
+import io.reactivex.disposables.*
+import io.reactivex.plugins.*
 import kotlinx.coroutines.*
-import java.util.concurrent.TimeUnit
-import kotlin.coroutines.CoroutineContext
+import kotlinx.coroutines.channels.*
+import java.util.concurrent.*
+import kotlin.coroutines.*
 
 /**
  * Converts an instance of [Scheduler] to an implementation of [CoroutineDispatcher]
  * and provides native support of [delay] and [withTimeout].
  */
-public fun Scheduler.asCoroutineDispatcher(): SchedulerCoroutineDispatcher = SchedulerCoroutineDispatcher(this)
+public fun Scheduler.asCoroutineDispatcher(): CoroutineDispatcher =
+    if (this is DispatcherScheduler) {
+        dispatcher
+    } else {
+        SchedulerCoroutineDispatcher(this)
+    }
+
+/**
+ * Converts an instance of [CoroutineDispatcher] to an implementation of [Scheduler].
+ */
+public fun CoroutineDispatcher.asScheduler(): Scheduler =
+    if (this is SchedulerCoroutineDispatcher) {
+        scheduler
+    } else {
+        DispatcherScheduler(this)
+    }
+
+private class DispatcherScheduler(internal val dispatcher: CoroutineDispatcher) : Scheduler() {
+
+    val parentJob = SupervisorJob()
+    private val parentScope = CoroutineScope(parentJob)
+
+    override fun scheduleDirect(run: Runnable): Disposable {
+        val decoratedRun = RxJavaPlugins.onSchedule(run)
+        val worker = createWorker() as DispatcherWorker
```

Comment:
Can you just run the runnable directly on the dispatcher? It seems wasteful to create a worker with it's extra queuing machinery just to do this. The built-in rxjava schedulers seem to prefer running the task as directly as possible from these `scheduleDirect` methods, instead of delegating to their workers.

E.g.
```kotlin
val job = SupervisorJob(parentJob)
dispatcher.dispatch(job, decoratedRun)
// return the job as a Disposable
```

If that doesn't work, even just launching a coroutine to run this one runnable would probably be better than creating a whole worker.

Reply:
@zach-klippenstein that's not a bad idea, however it seems like it would break things for clients. ```createWorker``` is a public method so how should it behave if we don't use workers to run the tasks? Would we return null for ```createWorker```?

Reply:
No, `createWorker` is part of the Scheduler API, and your implementation of it seems reasonable (although I've never implemented a custom scheduler, I'm just comparing with Rx source). Take a look at the built-in schedulers' implementations to see what I mean. In fact the [documentation on `Scheduler` even says](https://github.com/ReactiveX/RxJava/blob/3.x/src/main/java/io/reactivex/rxjava3/core/Scheduler.java): `Implementors of this class are encouraged to provide a more efficient direct scheduling implementation to avoid the time and memory overhead of creating such {@code Worker}s for every task.`

`scheduleDirect` is used to schedule a single task, independently of any others. `createWorker` is used when multiple tasks need to be scheduled to run in order. I believe this is used to satisfy a lot of Rx contracts that require order to be preserved - eg `onNext`s always need to be executed in the order they were invoked in, so they use a worker. 

You can find the implementations of the built in schedulers here: https://github.com/ReactiveX/RxJava/tree/3.x/src/main/java/io/reactivex/rxjava3/internal/schedulers

(I've referenced rx3 docs/code here, but the contract is the same for rx2, you can look at the rx2 source in that same repo as well.)

Reply:
Ah I see! I read the docs on worker and scheduler but guess I misunderstood. Thanks for clearing that up to me and pointing to those examples! Yes i think I'll change my implementation.